File: packages/server/simple/rust/src/secrets.rs
use aes_gcm::{
    Aes256Gcm, Nonce,
    aead::{Aead, KeyInit, OsRng},
};
use rand::RngCore;
use serde_json::Value;
use std::path::PathBuf;

use crate::models::Secret;

pub const EVENT_SECRET_CREATE: &str = "secret.create";
pub const EVENT_SECRET_UPDATE: &str = "secret.update";
pub const EVENT_SECRET_DELETE: &str = "secret.delete";

pub fn get_or_create_secrets_key(data_dir: &PathBuf) -> Result<Vec<u8>, String> {
    if let Ok(env_key) = std::env::var("SIMPLE_SECRETS_KEY") {
        return hex::decode(&env_key).map_err(|e| format!("decode env key: {e}"));
    }
    let key_path = data_dir.join("secrets.key");
    if let Ok(data) = std::fs::read_to_string(&key_path) {
        return hex::decode(data.trim()).map_err(|e| format!("decode key file: {e}"));
    }
    let mut key = vec![0u8; 32];
    rand::rngs::OsRng.fill_bytes(&mut key);
    let hex_key = hex::encode(&key);
    std::fs::write(&key_path, &hex_key).map_err(|e| format!("write key file: {e}"))?;
    Ok(key)
}

pub fn encrypt_secret(key: &[u8], plaintext: &str) -> Result<String, String> {
    let cipher = Aes256Gcm::new_from_slice(key).map_err(|e| format!("new cipher: {e}"))?;
    let mut nonce_bytes = vec![0u8; 12];
    OsRng.fill_bytes(&mut nonce_bytes);
    let nonce = Nonce::from_slice(&nonce_bytes);
    let ciphertext = cipher
        .encrypt(nonce, plaintext.as_bytes())
        .map_err(|e| format!("encrypt: {e}"))?;
    Ok(hex::encode(nonce_bytes) + ":" + &hex::encode(ciphertext))
}

pub fn decrypt_secret(key: &[u8], encrypted: &str) -> Result<String, String> {
    let parts: Vec<&str> = encrypted.splitn(2, ':').collect();
    if parts.len() != 2 {
        return Err("invalid encrypted format".into());
    }
    let nonce_bytes = hex::decode(parts[0]).map_err(|e| format!("decode nonce: {e}"))?;
    let ciphertext = hex::decode(parts[1]).map_err(|e| format!("decode ciphertext: {e}"))?;
    let cipher = Aes256Gcm::new_from_slice(key).map_err(|e| format!("new cipher: {e}"))?;
    let nonce = Nonce::from_slice(&nonce_bytes);
    let plaintext = cipher
        .decrypt(nonce, ciphertext.as_slice())
        .map_err(|e| format!("decrypt: {e}"))?;
    Ok(String::from_utf8(plaintext).map_err(|e| format!("utf8: {e}"))?)
}

pub fn webhook_secret_data(secret: &Secret) -> Value {
    serde_json::json!({
        "secret": secret,
    })
}

File: packages/server/backbone/go/src/secrets.go
package main

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

const (
	EventSecretCreate = "secret.create"
	EventSecretUpdate = "secret.update"
	EventSecretDelete = "secret.delete"
)

func getOrCreateSecretsKey(dataDir string) ([]byte, error) {
	envKey := os.Getenv("BACKBONE_SECRETS_KEY")
	if envKey != "" {
		return hex.DecodeString(envKey)
	}
	keyPath := filepath.Join(dataDir, "secrets.key")
	if data, err := os.ReadFile(keyPath); err == nil {
		return hex.DecodeString(strings.TrimSpace(string(data)))
	}
	key := make([]byte, 32)
	if _, err := rand.Read(key); err != nil {
		return nil, fmt.Errorf("generate key: %w", err)
	}
	if err := os.WriteFile(keyPath, []byte(hex.EncodeToString(key)), 0600); err != nil {
		return nil, fmt.Errorf("write key file: %w", err)
	}
	return key, nil
}

func encryptSecret(key []byte, plaintext string) (string, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", fmt.Errorf("new cipher: %w", err)
	}
	aead, err := cipher.NewGCM(block)
	if err != nil {
		return "", fmt.Errorf("new gcm: %w", err)
	}
	nonce := make([]byte, aead.NonceSize())
	if _, err := rand.Read(nonce); err != nil {
		return "", fmt.Errorf("rand nonce: %w", err)
	}
	ciphertext := aead.Seal(nil, nonce, []byte(plaintext), nil)
	return hex.EncodeToString(nonce) + ":" + hex.EncodeToString(ciphertext), nil
}

func decryptSecret(key []byte, encrypted string) (string, error) {
	parts := strings.SplitN(encrypted, ":", 2)
	if len(parts) != 2 {
		return "", fmt.Errorf("invalid encrypted format")
	}
	nonce, err := hex.DecodeString(parts[0])
	if err != nil {
		return "", fmt.Errorf("decode nonce: %w", err)
	}
	ciphertext, err := hex.DecodeString(parts[1])
	if err != nil {
		return "", fmt.Errorf("decode ciphertext: %w", err)
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", fmt.Errorf("new cipher: %w", err)
	}
	aead, err := cipher.NewGCM(block)
	if err != nil {
		return "", fmt.Errorf("new gcm: %w", err)
	}
	plaintext, err := aead.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return "", fmt.Errorf("decrypt: %w", err)
	}
	return string(plaintext), nil
}

func webhookSecretData(secret *Secret) map[string]any {
	return map[string]any{
		"secret": secret,
	}
}
